### PR TITLE
Augmentations

### PR DESCRIPTION
**Pull request by jennyfolkesson**
Added support for 3D augmentations

Made (zscore) normalize for tiles a config variable

Added tests for Dataset class

### CODE REVIEW

**Comment by jennyfolkesson** [comment]
@anithapk @smguo Sorry for making commits after making a PR. I just wanted to clean up a little, I'm all done now. Thanks.

File: micro_dl/cli/train_script.py [hunk 1]
```diff
@@ -75,63 +74,34 @@ def create_datasets(df_meta,
...
-    )
-    test_dataset = BaseDataSet(
-        tile_dir=tile_dir,
-        input_fnames=test_metadata['fpaths_input'],
-        target_fnames=test_metadata['fpaths_target'],
-        batch_size=trainer_config['batch_size'],
-        model_task=model_task,
-    )
-    return train_dataset, val_dataset, test_dataset, split_samples
+    return all_datasets[0], all_datasets[1], all_datasets[2], split_samples
```
Comment (anithapk) [inline micro_dl/cli/train_script.py]: 🥇

File: micro_dl/input/dataset.py [hunk 2]
```diff
@@ -17,12 +17,9 @@ def __init__(self,
                  tile_dir,
                  input_fnames,
                  target_fnames,
+                 dataset_config,
```
Comment (anithapk) [inline micro_dl/input/dataset.py]: 💯

File: micro_dl/input/dataset.py [hunk 3]
```diff
@@ -69,21 +93,35 @@ def _augment_image(self, input_image, aug_idx):
          5 - rot 270
         :return np.array image after transformation is applied
         """
+        # We need to flip over different dimensions depending on data format
```
Comment (anithapk) [inline micro_dl/input/dataset.py]: I believe at this stage, we are augmenting either a 2D array (y-x) or 3D array (z-y-x), no channel axis available yet?

File: micro_dl/input/dataset.py [hunk 4]
```diff
@@ -69,21 +93,35 @@ def _augment_image(self, input_image, aug_idx):
...
+        if self.data_format == 'channels_first':
+            add_dim = 1
 
-        assert len(input_image.shape) == 2, 'current implementation works' \
-                                            'for 2D images only'
         if aug_idx == 0:
             return input_image
         elif aug_idx == 1:
-            trans_image = np.fliplr(input_image)
+            trans_image = np.flip(input_image, 1 + add_dim)
```
Comment (anithapk) [inline micro_dl/input/dataset.py]: # flip about axis=1 (which is row in numpy, hence about y)

Comment (anithapk) [reply micro_dl/input/dataset.py]: Could we please add these notes? (sorry, i usually get confused with rotation axes, hence trying to add these)

File: micro_dl/input/dataset.py [hunk 5]
```diff
@@ -69,21 +93,35 @@ def _augment_image(self, input_image, aug_idx):
...
         elif aug_idx == 2:
-            trans_image = np.flipud(input_image)
+            trans_image = np.flip(input_image, 0 + add_dim)
```
Comment (anithapk) [inline micro_dl/input/dataset.py]: ``` # flip about axis=0 (which is cols in numpy, hence about x)```

File: micro_dl/input/dataset.py [hunk 6]
```diff
@@ -69,21 +93,35 @@ def _augment_image(self, input_image, aug_idx):
...
         elif aug_idx == 3:
-            trans_image = np.rot90(input_image, 1)
+            trans_image = np.rot90(
```
Comment (anithapk) [inline micro_dl/input/dataset.py]: ``` # rot in plane defined by axis=(0, 1)```

File: micro_dl/input/dataset.py [hunk 7]
```diff
@@ -69,21 +93,35 @@ def _augment_image(self, input_image, aug_idx):
          5 - rot 270
         :return np.array image after transformation is applied
         """
+        # We need to flip over different dimensions depending on data format
+        add_dim = 0
+        if self.data_format == 'channels_first':
```
Comment (anithapk) [inline micro_dl/input/dataset.py]: may be: ```if len(input_image.shape)==3: ```

File: micro_dl/utils/train_utils.py [hunk 8]
```diff
@@ -65,20 +65,20 @@ def split_train_val_test(sample_set, train_ratio, test_ratio,
...
         num_val = max(num_val, 1)
         val_idx = np.random.choice(rem_set, num_val, replace=False)
-        split_idx['val'] = val_idx
+        split_idx['val'] = val_idx.tolist()
         rem_set = set(rem_set) - set(val_idx)
         rem_set = list(rem_set)
 
     train_idx = np.array(rem_set, dtype='int')
-    split_idx['train'] = train_idx
+    split_idx['train'] = train_idx.tolist()
```
Comment (anithapk) [inline micro_dl/utils/train_utils.py]: 👍

File: micro_dl/input/dataset.py (hunk 3 above)
Comment (jennyfolkesson) [reply micro_dl/input/dataset.py]: Yes, I'm holding off on dealing with channels for now. I'll clarify that it doesn't deal with channels in the docstring.
When you say axis 1-2 do you mean for channels first data?
Good point about flipping about z!

File: micro_dl/input/dataset.py (hunk 4 above)
Comment (jennyfolkesson) [reply micro_dl/input/dataset.py]: Sure!

File: micro_dl/input/dataset.py (hunk 3 above)
Comment (anithapk) [reply micro_dl/input/dataset.py]: Sorry about the confusion here, trying to separate z and channel axis! 1-2 here is w.r.to 3D array in z-y-x format.

File: micro_dl/input/dataset.py (hunk 7 above)
Comment (jennyfolkesson) [reply micro_dl/input/dataset.py]: Good ideay, but I think I changed the tiles to always have a singleton dimension, so len is always 3...

[the remark this answers is not in the document]
Comment (jennyfolkesson) [reply micro_dl/input/dataset.py]: There was an error in the test! So sorry for the confusion!

**Comment by jennyfolkesson** [comment]
@anithapk I thought I was being clever with dimension flipping for either data format, but as you point out in 1) the images are always channels first so I could have written augment_image for channels first only then flipped axes in case of channels_last after augmentations are done. :-D
